docs/excel/custom-functions-best-practices.md

```diff
@@ -0,0 +1,184 @@
+---
+ms.date: 09/18/2018
+description: Learn best practices and recommended patterns for Excel custom functions.
+title: Custom Functions' best practices
+---
+
+# Custom Functions' best practices
+
+This article covers some recommended patterns and solutions to common use cases with Excel custom functions.
+
+## Authentication
+
+Non-custom functions add-ins which have web views (i.e. display information in the taskpane) accomplish authentication via `Office.context.displayDialogAsync`. Custom functions differ from this approach. Because custom functions use a different runtime, they do not have access to the Office.context methods. Instead, they use methods available on the `OfficeRuntime` object, such as `OfficeRuntime.DisplayWebDialog()` and `OfficeRuntime.AsyncStorage.setItem(key, value)`.
+
+The following code sample shows how to store a token in AsyncStorage and display a dialog box which can indicate whether or not the user is authenticated.  
+
+```js
+// Get auth token before calling my service, a hypothetical API, which will deliver a stock price based on stock ticker string, such as "MSFT"
+async function getStock(ticker) {
+    const token = await getToken();
+    let data = await (await fetch(https://myservice.com/?token=token&ticker= + ticker).json());
+    return data.price;
+}
+
+async function getToken() {
+    if (_cachedToken) {
+        return _cachedToken;
+    } else {
+        return await getTokenViaDialog_AsPromise();
+    }
+}
+  
+// Function to display dialog window
+function getTokenViaDialog_AsPromise() {
+    return new Promise ((resolve, reject) => {
+        displayWebDialog("https://www.auth.com/", {
+           height: ’50’,
+           width: ’50%’,
+           hideTitle: true,
+           onMessage: (message, dialog) => {
+               let json = JSON.parse(message);
+                if (json.type === "token_succeeded") {
+                        resolve(json.value);
+                        dialog.closeDialog();
+                        return;
+                    }
+            // Otherwise, handle other messages.
+            },
+          onClose: () => reject("User closed dialog")
+        }).catch(e => reject(e));
+    });
+}
+```
+
+Looking for a sample of authentication not specific to Excel custom functions? See the article on [authorizing external services for add-ins which do not use the new JavaScript runtime](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/auth-external-add-ins).
+
+## Batching of API requests
+
+Instead of executing individual web requests, it is possible to batch requests using custom functions.
```

Comment:
This isn't remotely close to enough explanation for batching.

Reply:
Seems like a fair point. I can bulk up this section with more examples. I think I may have assumed that people are already familiar with the concept of batching and just needed a sample, but it would make sense to have more explanation around how to do it and what it is.

Reply:
I'm marking this as resolved because my plan is to take out batching and auth for now unless Reeza thinks otherwise.